src/deny.ts

```diff
@@ -3,36 +3,37 @@ import * as core from '@actions/core'
 
 export async function getDeniedChanges(
   changes: Change[],
-  deniedPackages: string[],
-  deniedGroups: string[]
+  deniedPackages: string[] = [],
+  deniedGroups: string[] = []
 ): Promise<Change[]> {
   const changesDenied: Change[] = []
 
-  let failed = false
+  let hasDeniedPackage = false
   for (const change of changes) {
-    change.name = change.name.toLowerCase()
-    const packageUrl = change.package_url.toLowerCase().split('@')[0]
+    const [packageName, packageVersion] = change.package_url
+      .toLowerCase()
+      .split('@')
 
-    if (deniedPackages) {
-      for (const denied of deniedPackages) {
-        if (packageUrl === denied.split('@')[0].toLowerCase()) {
-          changesDenied.push(change)
-          failed = true
-        }
+    for (const denied of deniedPackages) {
+      const [deniedName, deniedVersion] = denied.toLowerCase().split('@')
+      if (
+        (!deniedVersion || packageVersion === deniedVersion) &&
+        packageName === deniedName
+      ) {
+        changesDenied.push(change)
+        hasDeniedPackage = true
       }
     }
 
-    if (deniedGroups) {
-      for (const denied of deniedGroups) {
-        if (packageUrl.startsWith(denied.toLowerCase())) {
-          changesDenied.push(change)
-          failed = true
-        }
+    for (const denied of deniedGroups) {
+      if (packageName.startsWith(denied.toLowerCase())) {
```

Comment:
The use of `startsWith` here makes me a little nervous. For example, if someone wanted to deny [pkg:npm/j](https://www.npmjs.com/package/j), would that also end up denying `pkg:npm/jest`?

Reply:
Good catch! Matching on substrings like this could cause issues like in your example. I'm thinking to deny against exact namespace matches so there's no ambiguity.